e2e/setup.ts

```diff
@@ -0,0 +1,128 @@
+    }
+    const logFilePath = `${logFolderPath}/bitcoind.log`;
+
+    const config = yaml.load(readFileSync(configPath, 'utf8')) as Record<
+        string,
+        any
+    >;
+
+    const user = config.bitcoinCore.rpcUser;
+    const password = config.bitcoinCore.rpcPass;
+    const network = config.app.network;
+    const port = config.bitcoinCore.rpcPort;
+
+    try {
+        const docker = new Docker();
+        const imageName = 'btcpayserver/bitcoin:24.0.1-1';
+
+        const images = await docker.listImages();
+        const imageExists = images.some(
+            (image) => image.RepoTags && image.RepoTags.includes(imageName),
+        );
+
+        // Pull the image if it's not available
+        if (!imageExists) {
+            logger.log(`Image ${imageName} not found locally. Pulling...`);
+            const stream = await docker.pull(imageName);
+            await new Promise((resolve, reject) => {
+                docker.modem.followProgress(stream, (err, output) =>
+                    err ? reject(err) : resolve(output),
+                ),
+                    (event) => logger.log(event.status);
+            });
+        }
+
+        // Create and start the container
+        container = await docker.createContainer({
+            Image: imageName,
+            ExposedPorts: { [`${port}/tcp`]: {} }, // Expose the Bitcoin RPC and P2P ports
+            HostConfig: {
+                PortBindings: {
+                    [`${port}/tcp`]: [{ HostPort: `${port}` }],
+                },
+            },
+            Env: [
+                `BITCOIN_NETWORK=${network}`,
+                `BITCOIN_EXTRA_ARGS=server=1\n
+                    rest=1\n
+                    rpcbind=0.0.0.0:${port}\n
+                    rpcallowip=0.0.0.0/0\n
+                    rpcuser=${user}\n
+                    rpcpassword=${password}\n
+                    debug=0\n
+                    logips=1\n
+                    logtimemicros=1\n
+                    blockmintxfee=0\n
+                    deprecatedrpc=signrawtransaction\n
+                    listenonion=0\n
+                    fallbackfee=0.00001\n
+                    txindex=1`,
+            ],
+        });
+
+        logger.log('Starting bitcoind container...');
+        await container.start();
+        logger.log('bitcoind is now running inside Docker.');
+
+        // Pipe the container logs to the file
+        const logs = await container.logs({
+            follow: true,
+            stdout: true,
+            stderr: true,
+        });
+        logs.pipe(createWriteStream(logFilePath, { flags: 'w' }));
+
+        return container;
+    } catch (error) {
+        logger.error('Error starting bitcoind container:', error);
+        await container.remove({ v: true, force: true });
+        throw new Error(error);
+    }
+}
+
+async function setupTestApp(): Promise<INestApplication> {
+    process.env.NODE_ENV = 'e2e';
+    process.env.ROOT_DIR = join(__dirname, '..');
```

Comment:
this is a bad practice, environment variables should not be set programmatically